### PR TITLE
[WFLY-4880] Refactor the security manager permissions to be list

### PR DESCRIPTION
**Pull request by kabir**
attributes

Since this has only been in WF 9 and no product releases, I am skipping the transformers since they will be slightly complicated.

https://issues.jboss.org/browse/WFLY-4880

The new output is:

```
[standalone@localhost:9990 /] /subsystem=security-manager/deployment-permissions=default:read-resource
{
    "outcome" => "success",
        "result" => {
    "maximum-permissions" => [{"class" => "java.security.AllPermission"}],
    "minimum-permissions" => undefined
}
}
[standalone@localhost:9990 /] /subsystem=security-manager/deployment-permissions=default:read-resource-description
{
    "outcome" => "success",
        "result" => {
    "description" => "The permissions that are to be granted to deployments",
            "attributes" => {
        "maximum-permissions" => {
            "type" => LIST,
                    "description" => "The maximum set of permissions that can be granted to a deployment or jars.",
                    "expressions-allowed" => false,
                    "nillable" => true,
                    "allowed" => [
            "class",
                    "name",
                    "actions",
                    "module"
            ],
            "value-type" => {
                "class" => {
                    "type" => STRING,
                            "description" => "The permission fully-qualified class name",
                            "expressions-allowed" => true,
                            "nillable" => false,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                },
                "name" => {
                    "type" => STRING,
                            "description" => "The permission target name",
                            "expressions-allowed" => true,
                            "nillable" => true,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                },
                "actions" => {
                    "type" => STRING,
                            "description" => "The permission actions",
                            "expressions-allowed" => true,
                            "nillable" => true,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                },
                "module" => {
                    "type" => STRING,
                            "description" => "The module to load the permission from. If not specified it will be loaded using the class loader of the security manager subsystem.",
                            "expressions-allowed" => true,
                            "nillable" => true,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                }
            },
            "access-type" => "read-write",
                    "storage" => "configuration",
                    "restart-required" => "all-services"
        },
        "minimum-permissions" => {
            "type" => LIST,
                    "description" => "The minimum set of permissions to be granted to a deployment or jars.",
                    "expressions-allowed" => false,
                    "nillable" => true,
                    "allowed" => [
            "class",
                    "name",
                    "actions",
                    "module"
            ],
            "value-type" => {
                "class" => {
                    "type" => STRING,
                            "description" => "The permission fully-qualified class name",
                            "expressions-allowed" => true,
                            "nillable" => false,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                },
                "name" => {
                    "type" => STRING,
                            "description" => "The permission target name",
                            "expressions-allowed" => true,
                            "nillable" => true,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                },
                "actions" => {
                    "type" => STRING,
                            "description" => "The permission actions",
                            "expressions-allowed" => true,
                            "nillable" => true,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                },
                "module" => {
                    "type" => STRING,
                            "description" => "The module to load the permission from. If not specified it will be loaded using the class loader of the security manager subsystem.",
                            "expressions-allowed" => true,
                            "nillable" => true,
                            "min-length" => 1L,
                            "max-length" => 2147483647L
                }
            },
            "access-type" => "read-write",
                    "storage" => "configuration",
                    "restart-required" => "all-services"
        }
    },
    "operations" => undefined,
            "notifications" => undefined,
            "children" => {}
}
}
```

### CODE REVIEW

File: security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java [hunk 1]
```diff
@@ -40,27 +67,203 @@
...
+        return Collections.unmodifiableCollection(Arrays.asList(ATTRIBUTES));
     }
 
-    @Override
-    public List<? extends PersistentResourceDefinition> getChildren() {
-        return CHILDREN;
+    /**
+     * A {@link PersistentResourceXMLDescription} implementation that knows how to parse/write security permission elements.
+     *
+     * From what I (Kabir) can tell it isn't possible to parse a complex list attributes mapped by child xml elements, so I am doing it myself
```
Comment (ctomc) [inline security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java]: well you are almost right :) 
Support for that was just added https://github.com/wildfly/wildfly-core/pull/891

Comment (kabir) [inline security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java]: Please create a Jira so the conversion can be done after this exists in core

Comment (ctomc) [inline security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java]: Will do.

File: security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerExtension.java [hunk 2]
```diff
@@ -64,6 +65,9 @@ public void initialize(final ExtensionContext context) {
         final ManagementResourceRegistration registration = subsystem.registerSubsystemModel(SecurityManagerRootDefinition.INSTANCE);
         registration.registerOperationHandler(GenericSubsystemDescribeHandler.DEFINITION, GenericSubsystemDescribeHandler.INSTANCE, false);
         subsystem.registerXMLElementWriter(SecurityManagerSubsystemParser.INSTANCE);
+
+        //Since version 1.0.0 of this subsystem has not been part of any product release I am not registering
+        //any transformers for it
```
Comment (bstansberry) [inline security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerExtension.java]: For the record: I discussed this with Jason and he agreed this is fine. We can look into transforming to WildFly 9.x if user demand arises.

I don't foresee ever transforming to 8.x.

**Comment by bstansberry** [comment]
@kabir :+1: 

Please feel free to merge if the testsuite run looks good. If you don't get to it and the tests are good I'll put the read to merge label on it so it can get merged whenever.

**Pull request merged**